src/shims/foreign_items.rs

```diff
@@ -483,6 +483,40 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 };
             }
 
+            // Allocates memory directly from miri.
+            "miri_alloc" => {
+                let [size, align] = this.check_shim(abi, Abi::Rust, link_name, args)?;
+                let size = this.read_scalar(size)?.to_machine_usize(this)?;
+                let align = this.read_scalar(align)?.to_machine_usize(this)?;
+
+                Self::check_alloc_request(size, align)?;
+
+                let ptr = this.allocate_ptr(
+                    Size::from_bytes(size),
+                    Align::from_bytes(align).unwrap(),
+                    MiriMemoryKind::Rust.into(),
+                )?;
+
+                this.write_pointer(ptr, dest)?;
+
+                return Ok(EmulateByNameResult::NeedsJumping);
```

Comment:
this shares a lot of logic with the `__*` code paths. I think we could share the logic with some well placed closures and conditions.